src/libraries/System.Text.RegularExpressions/tests/Regex.KnownPattern.Tests.cs

```diff
@@ -137,13 +142,759 @@ string DomainMapper(Match match)
+        }
+
+        // https://docs.microsoft.com/en-us/dotnet/standard/base-types/backtracking-in-regular-expressions#backtracking-with-optional-quantifiers-or-alternation-constructs
+        [Theory]
+        [InlineData(RegexOptions.None)]
+        [InlineData(RegexOptions.Compiled)]
+        public void Docs_Backtracking_WithOptionalQuantifiersOrAlternationConstructs(RegexOptions options)
+        {
+            const string Pattern = ".*(es)";
+            const string Input = "Essential services are provided by regular expressions.";
+
+            Match m = Regex.Match(Input, Pattern, RegexOptions.IgnoreCase | options);
+            Assert.True(m.Success);
+            Assert.Equal("Essential services are provided by regular expres", m.Value);
+            Assert.Equal(0, m.Index);
+            Assert.Equal(47, m.Groups[1].Index);
+
+            Assert.False(m.NextMatch().Success);
+        }
+
+        // https://docs.microsoft.com/en-us/dotnet/standard/base-types/backtracking-in-regular-expressions#nonbacktracking-subexpression
+        [Theory]
+        [InlineData(RegexOptions.None)]
+        [InlineData(RegexOptions.Compiled)]
+        public void Docs_Backtracking_WithNestedOptionalQuantifiers(RegexOptions options)
+        {
+            const string Input = "b51:4:1DB:9EE1:5:27d60:f44:D4:cd:E:5:0A5:4a:D24:41Ad:";
+            // Assert.False(Regex.IsMatch(Input, "^(([0-9a-fA-F]{1,4}:)*([0-9a-fA-F]{1,4}))*(::)$")); // takes too long due to backtracking
+            Assert.False(Regex.IsMatch(Input, "^((?>[0-9a-fA-F]{1,4}:)*(?>[0-9a-fA-F]{1,4}))*(::)$", options)); // non-backtracking
+        }
+
+        // https://docs.microsoft.com/en-us/dotnet/standard/base-types/backtracking-in-regular-expressions#lookbehind-assertions
+        [Theory]
+        [InlineData(RegexOptions.None)]
+        [InlineData(RegexOptions.Compiled)]
+        public void Docs_Backtracking_LookbehindAssertions(RegexOptions options)
+        {
+            const string Input = "test@contoso.com";
+
+            const string Pattern = @"^[0-9A-Z]([-.\w]*[0-9A-Z])?@";
+            Assert.True(Regex.IsMatch(Input, Pattern, RegexOptions.IgnoreCase | options));
+
+            const string BehindPattern = @"^[0-9A-Z][-.\w]*(?<=[0-9A-Z])@";
+            Assert.True(Regex.IsMatch(Input, BehindPattern, RegexOptions.IgnoreCase | options));
+        }
+
+        // https://docs.microsoft.com/en-us/dotnet/standard/base-types/backtracking-in-regular-expressions#lookahead-assertions
+        [Theory]
+        [InlineData(RegexOptions.None)]
+        [InlineData(RegexOptions.Compiled)]
+        public void Docs_Backtracking_LookaheadAssertions(RegexOptions options)
+        {
+            const string Input = "aaaaaaaaaaaaaaaaaaaaaa.";
+
+            //const string Pattern = @"^(([A-Z]\w*)+\.)*[A-Z]\w*$";
+            //Assert.False(Regex.IsMatch(Input, Pattern, RegexOptions.IgnoreCase)); // takes too long due to backtracking
```

Comment:
> // takes too long due to backtracking

That's interesting. Is this a candidate for optimization (rewriting the tree)? After that it seems we have a test ready to protect the optimization.

Reply:
> Is this a candidate for optimization (rewriting the tree)?

Into what?

Reply:
I hadn't figured that out yet -- was thinking 😄